### PR TITLE
Index nested loop join optimizer

### PR DESCRIPTION
**Pull request by andyfengHKU**
- Add isPrimaryKey check to PropertyExpression
- Make IndexScan a non-source operator
- Add index-nested-loop-join optimizer as a temporary solution before generic hash join is implemented.

### CODE REVIEW

File: src/binder/bind/bind_graph_pattern.cpp [hunk 1]
```diff
@@ -211,10 +211,16 @@ shared_ptr<NodeExpression> Binder::createQueryNode(const NodePattern& nodePatter
     for (auto tableID : tableIDs) {
         nodeTableSchemas.push_back(catalog.getReadOnlyVersion()->getNodeTableSchema(tableID));
     }
+    auto isSingleTable = nodeTableSchemas.size() == 1;
     for (auto& [propertyName, propertySchemas] :
         getNodePropertyNameAndPropertiesPairs(nodeTableSchemas)) {
+        auto isPrimaryKey = false;
```
Comment (ray6080) [inline src/binder/bind/bind_graph_pattern.cpp:217]: auto isPrimaryKey = isSingleTable && (nodeTableSchemas[0]->getPrimaryKey().propertyID == propertySchemas[0].propertyID)

File: src/include/binder/expression/property_expression.h [hunk 2]
```diff
@@ -40,6 +44,7 @@ class PropertyExpression : public Expression {
     }
 
 private:
+    bool isPrimaryKey_ = false;
```
Comment (ray6080) [inline src/include/binder/expression/property_expression.h:47]: Why do we need the `_` postfix?

File: src/include/optimizer/optimizer.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include "planner/logical_plan/logical_plan.h"
+
+namespace kuzu {
+namespace optimizer {
+
+class Optimizer {
+public:
+    Optimizer() = default;
```
Comment (ray6080) [inline src/include/optimizer/optimizer.h:10]: Do we need these default constructor and destructor?

File: src/include/optimizer/optimizer.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    ~Optimizer() = default;
+
+    void optimize(planner::LogicalPlan* plan);
```
Comment (ray6080) [inline src/include/optimizer/optimizer.h:13]: Can we make this function static?

File: src/include/optimizer/index_nested_loop_join_optimizer.h [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+// This is a temporary optimizer that rewrites
+//      Filter(a.ID=b.ID)
+//      CrossProduct                   into           IndexNestedLoopJoin(b)
+//   S(a)           S(b)                              S(a)
+// In the absense of a generic hash join operator.
+// We should merge this operator to filter push down + ASP when the generic hash join is
+// implemented.
+class IndexNestedLoopJoinOptimizer {
+public:
+    IndexNestedLoopJoinOptimizer() = default;
```
Comment (ray6080) [inline src/include/optimizer/index_nested_loop_join_optimizer.h:19]: Ditto. Do we need these default constructor and destructor?

File: src/include/optimizer/index_nested_loop_join_optimizer.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    ~IndexNestedLoopJoinOptimizer() = default;
+
+    shared_ptr<planner::LogicalOperator> rewrite(shared_ptr<planner::LogicalOperator> op);
```
Comment (ray6080) [inline src/include/optimizer/index_nested_loop_join_optimizer.h:22]: Why not set this function as static?

**Review by ray6080 — APPROVED** [review]

File: src/include/binder/expression/property_expression.h (hunk 2 above)
Comment (andyfengHKU) [reply src/include/binder/expression/property_expression.h:47]: because it conflicts with get function name

**Pull request merged**